packages/next/lib/page-loader.js

```diff
@@ -110,10 +123,30 @@ export default class PageLoader {
     }
   }
 
+  async prefetch (route) {
+    route = this.normalizeRoute(route)
+    const scriptRoute = route === '/' ? '/index.js' : `${route}.js`
+    if (this.prefetchCache.has(scriptRoute)) {
+      return
+    }
+    this.prefetchCache.add(scriptRoute)
+
+    const link = document.createElement('link')
+    // Feature detection is used to see if prefetch is supported, else fall back to preload
+    // Mainly this is for Safari
+    // https://caniuse.com/#feat=link-rel-prefetch
+    // https://caniuse.com/#feat=link-rel-preload
+    link.rel = supportsPrefetch ? 'prefetch' : 'preload'
```

Comment:
These aren't semantically the same. Prefetch should only be used for future navigations (as in full page loads). For sub-resources fetches (like this one), preload is the right one to use.

Prefetch may appear to work for client side navigations, but may lead to double-downloads if the prefetch hasn't finished by the time the actual request is made. That is because preload hits the memory cache (which de-dupes requests) while prefetch is only de-duped once the response made it into the disk cache.

Reply:
Hmm interesting, so you're saying we should use `preload` instead, but that triggers a warning if not used within 3 seconds, which is likely to cause  confusion to users of Next.

Reply:
Yeah, there really needs to be a "I know what I'm doing" flag to remove that warning. Very similar things happen with lazy loading resources before rendering where render-time depends on scroll.

CC @yoavweiss @igrigorik for actually authoritative advice.

Reply:
Created this PR: https://github.com/zeit/next.js/pull/5744

Will merge it once @yoavweiss / @igrigorik get back about this.

Reply:
Indeed `preload` is a better choice for same page resources. On top of that, the future of `prefetch` for subresources is a [bit murky](https://github.com/w3c/resource-hints/issues/82#issuecomment-425633915).